File: extracted_code/python/tts speak ssml TextToSpeech__snippet-012.py
package com.exocortex.neuroacoustic

import android.Manifest
import android.app.Service
import android.content.Intent
import android.content.pm.PackageManager
import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.AudioTrack
import android.media.MediaRecorder
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import be.tarsos.dsp.AudioDispatcher
import be.tarsos.dsp.AudioEvent
import be.tarsos.dsp.AudioProcessor
import be.tarsos.dsp.pitch.PitchDetectionHandler
import be.tarsos.dsp.pitch.PitchDetectionResult
import be.tarsos.dsp.pitch.PitchProcessor
import be.tarsos.dsp.io.jvm.AudioDispatcherFactory
import be.tarsos.dsp.pitch.PitchShift
import be.tarsos.dsp.resample.RateTransposer
import be.tarsos.dsp.gain.GainProcessor
import com.k2fsa.sherpa.onnx.TtsConfig
import com.k2fsa.sherpa.onnx.OfflineTts
import org.apache.commons.math3.linear.Array2DRowRealMatrix
import org.apache.commons.math3.linear.RealMatrix
import org.apache.commons.math3.ode.FirstOrderDifferentialEquations
import org.apache.commons.math3.ode.nonstiff.DormandPrince853Integrator
import org.kaldi.Model
import org.kaldi.RecognitionListener
import org.kaldi.Vosk
import com.google.mediapipe.tasks.genai.llminference.LlmInference
import com.google.mediapipe.tasks.genai.llminference.LlmInferenceOptions
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.util.Locale
import java.util.concurrent.Executors
import kotlin.math.exp
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt
import kotlin.random.Random

class MainActivity : AppCompatActivity() {
    private lateinit var statusText: TextView
    private lateinit var startButton: Button
    private val RECORD_REQUEST_CODE = 101
    private var isServiceRunning = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        statusText = findViewById(R.id.id_status)
        startButton = findViewById(R.id.id_start)

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.RECORD_AUDIO), RECORD_REQUEST_CODE)
        } else {
            initSystem()
        }

        startButton.setOnClickListener {
            if (!isServiceRunning) {
                startService(Intent(this, ExocortexService::class.java))
                startButton.text = "Stop Listening"
                isServiceRunning = true
            } else {
                stopService(Intent(this, ExocortexService::class.java))
                startButton.text = "Start Listening"
                isServiceRunning = false
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == RECORD_REQUEST_CODE && grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            initSystem()
        } else {
            Toast.makeText(this, "Audio permission denied", Toast.LENGTH_SHORT).show()
        }
    }

    private fun initSystem() {
        // Check models
        val ttsModelFile = File(filesDir, "vits-piper-en_US-amy-medium.onnx")
        if (!ttsModelFile.exists()) {
            try {
                assets.open("vits-piper-en_US-amy-medium.onnx").use { input ->
                    FileOutputStream(ttsModelFile).use { output ->
                        input.copyTo(output)
                    }
                }
            } catch (e: Exception) {
                statusText.text = "TTS model copy failed: ${e.message}"
                return
            }
        }
        val llmModelFile = File(filesDir, "gemma-1.1-2b-it-q4f16.task")
        if (!llmModelFile.exists()) {
            statusText.text = "LLM model missing. Download Gemma and place in app files."
            return
        }
        statusText.text = "System Initialized"
    }

    override fun onDestroy() {
        super.onDestroy()
    }

File: extracted_code/java/tts speak ssml TextToSpeech__snippet-012.java
class NeuroAcousticMirror(private val context: Context) {
    private val vadSilenceThresholdMs = 1200L
    private val sampleRate = 16000
    private val bufferSize = AudioRecord.getMinBufferSize(sampleRate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT) * 4
    private var audioRecord: AudioRecord? = null
    private var voskModel: Model? = null
    private val audioBuffer = ByteArrayOutputStream()
    private var offlineTts: OfflineTts? = null

    init {
        Vosk.init(context)
        val modelPath = File(context.filesDir, "vosk-model-en-us-0.22")
        if (!modelPath.exists()) {
            Toast.makeText(context, "Vosk model missing", Toast.LENGTH_SHORT).show()
        } else {
            voskModel = Model(modelPath.absolutePath)
        }

        // Init Sherpa TTS
        val ttsModelPath = File(context.filesDir, "vits-piper-en_US-amy-medium.onnx").absolutePath
        val config = TtsConfig(model = ttsModelPath, numThreads = 1, debug = false)
        offlineTts = OfflineTts(config)
    }

    fun listenAndProcess(callback: (String, Prosody) -> Unit) {
        if (voskModel == null || ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) return

        audioRecord = AudioRecord(MediaRecorder.AudioSource.MIC, sampleRate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT, bufferSize)
        audioRecord?.startRecording()
        audioBuffer.reset()

        val recognizer = Vosk.createRecognizer(voskModel, sampleRate.toFloat())
        recognizer.setRecognitionListener(object : RecognitionListener {
            override fun onPartialResult(hypothesis: String?) {}
            override fun onResult(hypothesis: String?) {
                val rawText = hypothesis?.let { parseJsonResult(it) } ?: return
                val prosody = extractProsody(audioBuffer.toByteArray(), rawText)
                val correctedText = correctToFirstPerson(rawText)
                speakCorrectedText(correctedText, prosody)
                callback(correctedText, prosody)
            }
            override fun onFinalResult(hypothesis: String?) {}
            override fun onError(e: Exception?) { Log.e("Mirror", "Recognition error: ${e?.message}") }
            override fun onTimeout() {}
        })

        val buffer = ByteArray(bufferSize)
        var silenceStart = 0L
        val startTime = System.currentTimeMillis()
        while (true) {
            val read = audioRecord?.read(buffer, 0, buffer.size) ?: 0
            if (read > 0) {
                audioBuffer.write(buffer, 0, read)
                val shortBuffer = ShortArray(read / 2)
                java.nio.ByteBuffer.wrap(buffer, 0, read).asShortBuffer().get(shortBuffer)
                recognizer.acceptWaveForm(shortBuffer, shortBuffer.size)
                val energy = calculateRMS(shortBuffer)
                if (energy < 0.01) {
                    if (silenceStart = 0L) silenceStart = System.currentTimeMillis()
                    if (System.currentTimeMillis() - silenceStart > vadSilenceThresholdMs && System.currentTimeMillis() - startTime > 1000) {
                        recognizer.result
                        break
                    }
                } else {
                    silenceStart = 0L
                }
            } else if (read < 0) {
                break
            }
        }

        audioRecord?.stop()
        audioRecord?.release()
    }

    private fun parseJsonResult(json: String): String {
        val textStart = json.indexOf("\"text\" : \"") + 10
        val textEnd = json.lastIndexOf("\"")
        return if (textStart > 9 && textEnd > textStart) json.substring(textStart, textEnd).trim() else ""
    }

    private fun calculateRMS(buffer: ShortArray): Double {
        var sum = 0.0
        buffer.forEach { sum += (it / 32768.0).pow(2) }
        return if (buffer.isNotEmpty()) sqrt(sum / buffer.size) else 0.0
    }

    private fun correctToFirstPerson(text: String): String {
        if (text.isEmpty()) return ""
        var corrected = text.replace(Regex("\\b(you|he|she|they)\\b", RegexOption.IGNORE_CASE), "I")
            .replace(Regex("\\b(your|his|her|their)\\b", RegexOption.IGNORE_CASE), "my")
        if (!corrected.matches(Regex("I .*", RegexOption.IGNORE_CASE))) {
            corrected = "I " + corrected.replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() }
        }
        return corrected
    }

    private fun extractProsody(audioBytes: ByteArray, text: String): Prosody {
        if (audioBytes.isEmpty()) return Prosody(120.0, 1.0, 0.5, 0.5, 20.0)

        val dispatcher = AudioDispatcherFactory.fromByteArray(audioBytes, sampleRate, 1024, 512)
        val pitches = mutableListOf<Float>()
        dispatcher.addAudioProcessor(PitchProcessor(PitchProcessor.PitchEstimationAlgorithm.YIN, sampleRate.toFloat(), 1024, object : PitchDetectionHandler {
            override fun handlePitch(result: PitchDetectionResult, event: AudioEvent) {
                if (result.pitch > 0) pitches.add(result.pitch)
            }
        }))
        dispatcher.run()
        val avgPitch = if (pitches.isNotEmpty()) pitches.average() else 120.0
        val pitchVariance = if (pitches.size > 1) pitches.map { (it - avgPitch).pow(2) }.average() else 0.0

        val durationSec = audioBytes.size / (sampleRate * 2).toDouble()
        val wordCount = text.split("\\s+".toRegex()).size.toDouble()
        val rate = if (durationSec > 0) (wordCount / durationSec) / 1.5 else 1.0

        val shortArray = ShortArray(audioBytes.size / 2)
        java.nio.ByteBuffer.wrap(audioBytes).asShortBuffer().get(shortArray)
        val volume = calculateRMS(shortArray)

        val arousal = ((avgPitch / 150.0) + (pitchVariance / 50.0) + (volume * 5.0)) / 3.0.coerceIn(0.0, 1.0)

        return Prosody(avgPitch, rate.coerceIn(0.5, 2.0), arousal, volume.coerceIn(0.0, 1.0), pitchVariance)
    }

    private fun speakCorrectedText(text: String, prosody: Prosody) {
        if (text.isEmpty() || offlineTts == null) return

        // Synthesize with default
        val generatedAudio = offlineTts!!.generate(text, speed = 1.0f, speakerId = 0) // Assume model supports speed and sid

        var samples = generatedAudio.samples // Assume float[] 
        val ttsSampleRate = generatedAudio.sampleRate // Assume 16000 or 22050

        // Post-process for prosody transfer
        val dispatcher = AudioDispatcherFactory.fromFloatArray(samples, ttsSampleRate, 1024, 512)

        // Adjust rate (time stretch)
        val rateTransposer = RateTransposer(prosody.rate)
        dispatcher.addAudioProcessor(rateTransposer)

        // Adjust pitch
        val pitchShiftFactor = prosody.pitch / 120.0 // Assume base pitch 120Hz
        val pitchShift = PitchShift(pitchShiftFactor, ttsSampleRate.toFloat())
        dispatcher.addAudioProcessor(pitchShift)

        // Adjust volume
        val gain = prosody.volume * 2.0 // Normalize to gain factor
        val gainProcessor = GainProcessor(gain)
        dispatcher.addAudioProcessor(gainProcessor)

        val processedBuffer = FloatArray(samples.size * 2) // Extra space
        var index = 0
        dispatcher.addAudioProcessor(object : AudioProcessor {
            override fun process(audioEvent: AudioEvent?): Boolean {
                val buffer = audioEvent!!.floatBuffer
                buffer.copyInto(processedBuffer, index)
                index += buffer.size
                return true
            }
            override fun processingFinished() {}
        })
        dispatcher.run()

        val processedSamples = processedBuffer.copyOf(index)

        // Play the processed audio
        playAudio(processedSamples, ttsSampleRate)
    }

    private fun playAudio(samples: FloatArray, sampleRate: Int) {
        val bufferSize = AudioTrack.getMinBufferSize(sampleRate, AudioFormat.CHANNEL_OUT_MONO, AudioFormat.ENCODING_PCM_FLOAT)
        val audioTrack = AudioTrack.Builder()
            .setAudioAttributes(AudioAttributes.Builder().setUsage(AudioAttributes.USAGE_MEDIA).setContentType(AudioAttributes.CONTENT_TYPE_SPEECH).build())
            .setAudioFormat(AudioFormat.Builder().setSampleRate(sampleRate).setChannelMask(AudioFormat.CHANNEL_OUT_MONO).setEncoding(AudioFormat.ENCODING_PCM_FLOAT).build())
            .setBufferSizeInBytes(bufferSize)
            .setTransferMode(AudioTrack.MODE_STREAM)
            .build()

        audioTrack.play()
        audioTrack.write(samples, 0, samples.size, AudioTrack.WRITE_BLOCKING)
        audioTrack.stop()
        audioTrack.release()
    }
